File: Packages/RecordingKit/Sources/RecordingKit/RecorderState.swift
import Foundation

/// Pure recording state machine. `recording`/`paused` carry the start time and
/// the total time spent paused so the elapsed timer can exclude pauses.
public enum RecorderState: Equatable {
    case idle
    case armed                                                   // record strip showing
    case recording(started: Date, accumulatedPause: TimeInterval)
    case paused(started: Date, accumulatedPause: TimeInterval, since: Date)
    case finishing                                              // writer finalizing — new commands rejected

    public enum Event: Equatable {
        case arm                     // show the strip
        case begin(Date)             // capture started
        case pause(Date)             // recording → paused at this time
        case resume(Date)            // paused → recording at this time
        case finish                  // stop requested
        case reset                   // back to idle (finalized or cancelled)
    }

    /// Applies `event` if legal; returns whether the state changed.
    @discardableResult
    public mutating func transition(_ event: Event) -> Bool {
        switch (self, event) {
        case (.idle, .arm):                       self = .armed
        case (.armed, .begin(let date)):          self = .recording(started: date, accumulatedPause: 0)
        case (.armed, .reset):                    self = .idle
        case (.recording(let started, let acc), .pause(let at)):
            self = .paused(started: started, accumulatedPause: acc, since: at)
        case (.paused(let started, let acc, let since), .resume(let at)):
            self = .recording(started: started, accumulatedPause: acc + at.timeIntervalSince(since))
        case (.recording, .finish):               self = .finishing
        case (.paused, .finish):                  self = .finishing
        case (.finishing, .reset):                self = .idle
        default:                                  return false
        }
        return true
    }

    /// "m:ss" while recording (paused time excluded); "Paused · m:ss" (frozen)
    /// while paused; nil otherwise.
    public func elapsedString(now: Date) -> String? {
        switch self {
        case .recording(let started, let acc):
            let secs = max(0, Int(now.timeIntervalSince(started) - acc))
            return "\(secs / 60):" + String(format: "%02d", secs % 60)
        case .paused(let started, let acc, let since):
            let secs = max(0, Int(since.timeIntervalSince(started) - acc))
            return "Paused · \(secs / 60):" + String(format: "%02d", secs % 60)
        default:
            return nil
        }
    }
}

File: windows/src/BetterScreenshot.Recording/RecorderState.cs
namespace BetterScreenshot.Recording;

public enum RecorderPhase { Idle, Armed, Recording, Paused, Finishing }

public enum RecorderEvent { Arm, Begin, Pause, Resume, Finish, Reset }

/// <summary>
/// Pure recording state machine: Idle → Armed → Recording ⇄ Paused → Finishing → Idle. Tracks the start time
/// and accumulated pause duration so elapsed time excludes pauses. Illegal transitions are rejected (return false).
/// </summary>
public struct RecorderState : IEquatable<RecorderState>
{
    public RecorderPhase Phase { get; private set; }
    public DateTime Started { get; private set; }
    public TimeSpan AccumulatedPause { get; private set; }
    public DateTime PausedSince { get; private set; }

    public static RecorderState Idle => new() { Phase = RecorderPhase.Idle };

    /// <summary>Applies an event. Returns true if the transition was legal (and applied), false otherwise.</summary>
    public bool Transition(RecorderEvent ev, DateTime at = default)
    {
        if (ev == RecorderEvent.Reset)
        {
            this = Idle;
            return true;
        }

        switch (Phase)
        {
            case RecorderPhase.Idle when ev == RecorderEvent.Arm:
                Phase = RecorderPhase.Armed;
                return true;
            case RecorderPhase.Armed when ev == RecorderEvent.Begin:
                Phase = RecorderPhase.Recording;
                Started = at;
                AccumulatedPause = TimeSpan.Zero;
                return true;
            case RecorderPhase.Recording when ev == RecorderEvent.Pause:
                Phase = RecorderPhase.Paused;
                PausedSince = at;
                return true;
            case RecorderPhase.Recording when ev == RecorderEvent.Finish:
                Phase = RecorderPhase.Finishing;
                return true;
            case RecorderPhase.Paused when ev == RecorderEvent.Resume:
                AccumulatedPause += at - PausedSince;
                Phase = RecorderPhase.Recording;
                return true;
            case RecorderPhase.Paused when ev == RecorderEvent.Finish:
                Phase = RecorderPhase.Finishing;
                return true;
            default:
                return false;
        }
    }

    /// <summary>"m:ss" while recording; "Paused · m:ss" (frozen) while paused; null otherwise.</summary>
    public string? ElapsedString(DateTime now)
    {
        return Phase switch
        {
            RecorderPhase.Recording => Format(now - Started - AccumulatedPause),
            RecorderPhase.Paused => "Paused · " + Format(PausedSince - Started - AccumulatedPause),
            _ => null,
        };
    }

    private static string Format(TimeSpan t)
    {
        if (t < TimeSpan.Zero) t = TimeSpan.Zero;
        int total = (int)t.TotalSeconds;
        return $"{total / 60}:{total % 60:D2}";
    }

    public bool Equals(RecorderState other) =>
        Phase == other.Phase && Started == other.Started &&
        AccumulatedPause == other.AccumulatedPause && PausedSince == other.PausedSince;

    public override bool Equals(object? obj) => obj is RecorderState s && Equals(s);
    public override int GetHashCode() => HashCode.Combine(Phase, Started, AccumulatedPause, PausedSince);
}
